Redesign info overlay

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
Rebased the Work by @Taiki-Choda in #23547.

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Before:
![image](https://github.com/zulip/zulip/assets/64723994/4cb3d93b-7d8e-45fb-ad69-2ec020a6f667)

After:
![image](https://github.com/zulip/zulip/assets/64723994/533f9598-c962-4389-bd70-240038049865)

Before:
![image](https://github.com/zulip/zulip/assets/64723994/ab9ab7c1-d2d5-4e12-ac0a-774e103ca919)

After:
![image](https://github.com/zulip/zulip/assets/64723994/7531fe0d-93f5-469e-99cb-fada8e467b2b)

Before:
![image](https://github.com/zulip/zulip/assets/64723994/2e9d7c4c-81af-46a6-b004-5e1fd6638649)

After:
![image](https://github.com/zulip/zulip/assets/64723994/ef1026b7-46a2-4726-9c78-096ee407d2a8)

Before:
![image](https://github.com/zulip/zulip/assets/64723994/db837c1f-0168-4822-8660-c7b76404a745)

After:
![image](https://github.com/zulip/zulip/assets/64723994/49c0898c-86f2-4636-adac-281d9c4783c4)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@zulipbot add "redesign" "buddy review" "QA needed"

prakhar1144 reviewed (commented):
@Ddharmani3 Thanks for rebasing!

Review comments:
* We should add **Taiki-Choda** as a co-author in the commits, if the commit re-uses some of the work done by him.
* Please compare the following with the existing design in CZO

    <details>
    <summary>
    <b>Incorrect keyboard shortcut for "Zoom to message in conversation context" and after that</b>
    </summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/64a376ea-62ff-474b-9e8a-fe94f0fc81ec"></img>
    </details>
    <details>
    <summary>
    <b>Missing Horizontal line</b>
    </summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/ae9164d9-f99b-4480-b0cc-f63cb949b680"></img>
    </details>
    <details>
    <summary>
    <b>Uneven Top and Bottom spacing</b>
    </summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/ac0ef132-4cc0-4700-9c48-28cbc9851bbc"></img>
    </details>
    <details>
    <summary>
    <b>Should we keep 'Effect' column a bit more right aligned?</b>
    </summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/3490539f-c181-4501-9519-3085316013e5"></img>
    </details>

karlstolley commented:
@Ddharmani3 if you could label your screenshots "Before" and "After" (and include both types), that would be super helpful :)

Ddharmani3 commented:
> @Ddharmani3 if you could label your screenshots "Before" and "After" (and include both types), that would be super helpful :)

@karlstolley , Updated the PR description to add before images too.
Also, the PR is updated to address all the feedback by @prakhar1144.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
         position: relative;
-        top: calc((30vh - 50px) / 2);
-        border-radius: 4px;
+        top: unset;
```
Again, because you're using Grid, you can strike a whole bunch of these properties entirely. `position: relative` doesn't lijkely need to happen, now that `.overlay-content` is a grid item. That means you don't even need a `top` declaration. `margin: 0 auto` can disappear to.

One thing I like to do when I'm working with CSS is to use my browsers inspector and just switch off different CSS properties. If there's no effect from switching the property off, I'll just delete it in my source code, and then spend some time browsing around to see if any distant elements are affected.

For example:

| Original | Switched off |
| --- | --- |
| <img width="995" alt="initial-values" src="https://github.com/zulip/zulip/assets/170719/fe7f7783-cbde-455d-9aa3-97929f1670bb"> | <img width="995" alt="switched-off" src="https://github.com/zulip/zulip/assets/170719/d67d5d0e-3ec5-42b9-9753-cf77622e4bba"> |

The overlay seems unaffected by those properties being switched off. So I'd delete them.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
-        border-bottom: 1px solid hsl(0deg 0% 0% / 20%);
+        position: relative;
+        display: flex;
+        flex-flow: row nowrap;
```
As above, I'd spend time seeing if there are additional properties here that don't need to be declared. Some might have no effect (again, there's typically no need for relative positioning in a grid context; that means you can lose `z-index` too), while some are defaults: `row nowrap` is [the default for `flex-flow`](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-flow#formal_definition). You get them just by setting `display: flex` on an item. Nice, right?

The only time you want to declare default values like that is if you're resetting from another, competing style declaration elsewhere (in which case you'd definitely want to put in a comment saying which style, and why).

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
-            color: hsl(0deg 0% 67%);
-            font-weight: 600;
-            margin: 0 15px;
+            display: inline-block;
```
So in this case, `.exit` is a flex item of `.overlay-tabs`. And because it's a flex item by itself, there's no need to tell it to display inline. If we just let flexbox do its thing, you can actually ditch almost all of these properties! I'll let you try this yourself to see the result, but there's no change:

<img width="300" alt="Screenshot 2023-06-16 at 11 16 56 AM" src="https://github.com/zulip/zulip/assets/170719/b86af47c-e6ba-4c9f-a86e-7b80225866ac">

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
-    .overlay-modal {
-        padding-bottom: 10px;
+    .modal-body {
+        height: 75vh;
```
Here again, the only property you really need to set is the `height`! Elements don't have a `max-height` unless you set one elsewhere, so there's no need to unset it. Center-aligning the text has no effect, yet there's a whole bunch of child elements (not of your PR's creation) that have `text-align: left` defined. So you could imagine striking the `text-align` property here, and probably being able to remove the default `text-align: left` from all those child elements (in another PR :)

Ddharmani3 replied:
Everything you mentioned in this comment was useless as you suggested but `max-height: unset` is needed as it was getting affected by the `max-height` property of `.modal-body` in `modal.css` and `bootstrap.css` I have also commented the same in code.

karlstolley replied:
Awesome! You'll save some future contributor a lot of headaches with comments like that. Thank you :)

karlstolley replied:
Yeah, definitely a follow-up item.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
+
+    & kbd {
+        font-family: "Source Sans 3", sans-serif;
+        font-feature-settings: "ss01" on;
```
Be sure to leave a comment about what `font-feature-settings` you're enabling here. `ss01` likely refers to some stylistic alternates, so it would be good to know which ones :)

Ddharmani3 replied:
This comment was added by @Taiki-Choda in the original PR. I did some research and discovered that it provides alternate glyphs for certain characters (different styling/variants).

Upon comparing the differences between using the `font-feature-settings` property and not using it, I noticed that it only impacts the character I, showcasing a distinct variant:

Without `font-feature-settings`: 
![image](https://github.com/zulip/zulip/assets/64723994/9138a34e-f9a9-43e4-bb08-0d25f17108b4)

With font-feature-settings::
![image](https://github.com/zulip/zulip/assets/64723994/09212faa-f18f-4981-a523-c0e53bb2404f)

The `I` with `font-feature-settings: "ss01" on;` was looking better so I kept it :)

karlstolley replied:
Ah, that's awesome. That alternate glyph makes a big difference (and I agree the version with the bars on the I should be kept), so it'll be very useful to have the comment explaining this in the code.

Ddharmani3 replied:
Added the same content that was in tooltips.css for hotkeys hints:
![image](https://github.com/zulip/zulip/assets/64723994/0df35c26-7bb0-4dba-81e2-3cf4d050df87)

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
 
-            & th {
-                font-weight: 400;
+            &:not(:last-child) {
```
Using flexbox would also mean you wouldn't have to use this complicated not/first-child selector to disallow right margin!

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
+    }
+
+    .concurrent-hotkeys {
+        display: inline-block;
```
(Though is looks like you'd need to make some adjustments to how hotkeys are presented under "The Basics" section if you move to flexbox; that would probably mean adjusting some HTML.)

Ddharmani3 replied:
I changed it to flex as you suggested and it is cleaner/less complex as not/first/last-child are now not required. I think you were suggesting an HTML change because the shortcuts with a `or` in between were extending to the next line and breaking CSS, right? ( like this: 
![image](https://github.com/zulip/zulip/assets/64723994/69e40bcf-1538-42f9-8725-e2c9f58ebe09))

I fixed this by using `display: inline-flex` and It is not breaking CSS and is giving the same design as before, can you check if this is the right approach?

karlstolley replied:
Yeah, in this case, `inline-flex` is probably the better approach. The issue seems to be the text nodes that we get with things like "or" that don't have markup of their own.

Ddharmani3 commented:
> Hi @Ddharmani3 -- I left a TON of comments for you :) please don't be discouraged by them. Happy to talk about any of this if it would be helpful, either here or on CZO.

@karlstolley, Thank you for taking the time to review this PR❤️ and providing such valuable feedback with detailed explanations. I truly appreciate your support and am grateful to be a part of this wonderful community. It was not at all discouraging in fact it was insightful and motivated me to write more precise code in the future :)

I have rebased the work in this PR by @Taiki-Choda, and since the PR was in the "Integration review" stage, I haven't made significant changes. However, I will make sure to thoroughly review any future PRs that I rebase for such modifications.

I have addressed all of your feedback in the updated PR and have left some comments where necessary.

Regarding the first comment, I couldn't fully comprehend it. I will send you a direct message on CZO to discuss it further.

Once again, thank you for everything😊.

karlstolley commented:
Looks like some good work here, @Ddharmani3!

Ddharmani3 commented:
Added the comment for the alternate glyph of character `I`.
This PR is ready for review now.

Ddharmani3 commented:
@karlstolley rebased this PR as well. This PR is ready for your review :).

Ddharmani3 commented:
Resolved the conflicts. @karlstolley, can you please take one more look and, if you approve the PR, add the 'integration review' label?

karlstolley commented:
Okay @Ddharmani3. A few things here:

* I don't see any horizontal rules on the Message formatting or Filters tabs, in light or dark mode.
* The name of the Source Sans font family has changed; it's now `"Source Sans 3 VF"`

The text in the borders representing keys doesn't seem to be sitting quite right to me, but I think I've now been staring at it too long.

Ddharmani3 commented:
@karlstolley, Updated the PR to address feedback. The horizontal rules were maybe messed up in the last rebase fixed it. updated `Source Sans 3` to `Source Sans 3 VF`.

> The text in the borders representing keys doesn't seem to be sitting quite right to me, but I think I've now been staring at it too long.

You are not extremely wrong there is a slight misalignment, here's what it is:

After removing the margin and padding from KBD, Here's what it looks like in this PR vs. CZO:
![KBD in this PR](https://github.com/zulip/zulip/assets/64723994/0f636b86-9473-4cd0-863f-61eaa4df5ec8)
![KBD in CZO](https://github.com/zulip/zulip/assets/64723994/833d56d3-4355-4e1b-995f-51a9134d0bfa)

You can see that KBD in this PR is slightly on top aligned vertically, the reason for this is the `Source Sans 3 VF` font in this PR vs. `monospace` in CZO, here's how it looks like with monospace in this PR (centered aligned vertically):
![KBD with monospace font in this PR](https://github.com/zulip/zulip/assets/64723994/1132438d-18f9-4e80-b187-8094d797ddd2)

But `Source Sans 3 VF` looks better with the UI, your thoughts on this?

<details>
    <summary>Source Sans 3 VF vs. monospace UI</summary>

| Source Sans 3 VF | monospace |
| ----------------- | ---------- |
| ![image](https://github.com/zulip/zulip/assets/64723994/d1974c42-ce24-4843-b8e7-e096e2410b19) | ![image](https://github.com/zulip/zulip/assets/64723994/d4e902f9-fe23-4c3e-ac20-251e02afa1b8) |

</details>

karlstolley commented:
Got it. If Source Sans hasn't been settled on for this change, we should probably have a design discussion about it on CZO.

But if it has, it would be best to go ahead and tune the spacing for the new typeface.

timabbott commented:
For the keyboard pills, I think we should copy (or perhaps even via shared CSS classes) the design choices we have for key pills in tooltips when hovering elements:
* Not having `+` between keys pressed together, just put the key pills next to each other.
* Whatever font choices we made there.

I think `.tooltip-hotkey-hints` in `tooltips.css` is the main set of classes for this.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -1,108 +1,186 @@
-        /* because zoom breaks at 525px perhaps due to rounding errors, so add a
-           trivial amount of width so it doesn't break. */
-        width: 550px;
+        width: 90%;
```
@Ddharmani3 I think this is the only unresolved item in my feedback on this. Hopefully it's a quick fix.

Ddharmani3 committed:
informational_overlays: Redesign keyboard shortcuts modal.

This commit only redesign keyboard shortcuts modal to a modern design.
The new CSS introduced in this commit will also be used to redesign
markdown_help and search_operators modals in next commits.

Co-authored-by: Taiki-Choda <chotik013@gmail.com>

Ddharmani3 committed:
informational_overlays: Redesign .overlay-tabs.

Coverted `.overlay-tabs` to a flex. Removed border-bottom instead
used box-shadow for partition between .overlay-tabs and .overlay-body.
Changed overlay cross button to zulip-icon-close SVG.

Co-authored-by: Taiki-Choda <chotik013@gmail.com>

Ddharmani3 committed:
informational_overlays: Redesign search operators modal.

Co-authored-by: Taiki-Choda <chotik013@gmail.com>

Ddharmani3 committed:
informational_overlays: Redesign markdown help modal.

Co-authored-by: Taiki-Choda <chotik013@gmail.com>
